docs/tech-notes/index-backfill.md

```diff
@@ -0,0 +1,156 @@
+
+In this case, the index entry found by `InitPut` has never been written to.
+Since we know that all operations on the table at or after `tScan` cause updates
+to the index, an absent index entry indicates that the primary table entry has
+not been updated since `tScan` and is therefore safe to echo to the index.
+
+For example, when the backfiller encounters the pair `1` `a`, the corresponding
+index entry is empty, so it proceeds to write.
+
+| `k` | `v` | `t` (MVCC timestamp) | `idx_k` (`v`) | `idx_v` (`k`) | `idx_t` (idx MVCC timestamp) |
+|-----|-----|----------------------|---------------|---------------|------------------------------|
+| `1` | `a` | 0                    | `a`           | `1`           | `after tScan`                |
+| `2` | `b` | `tWrite+1`           | `b`           | `2`           | `tWrite+1`                   |
+| `3` | `c` | 0                    | `c`           | `(tombstone)` | `tScan+1`                    |
+| `3` | `d` | `tScan+1`            | `d`           | `3`           | `tScan+1`                    |
+| `4` | `e` | 0                    |               |               |                              |
+| `5` | `e` | `tScan+1`            | `e`           | `5`           | `tScan+1`                    |
+| `6` | `f` | 0                    |               |               |                              |
+| `6` | ` ` | `tScan+1`            | `f`           | `(tombstone)` | `tScan+1`                    |
+| `7` | `g` | `0`                  |               |               |                              |
+| `8` | `g` | `tScan+1`            | `g`           | `8`           | `tScan+1`                    |
+
+### The `InitPut` sees a non-empty entry with identical value bytes
+
+In this case, the index entry found by `InitPut` has a value identical to the
+key of the table. There are just two ways this could happen, both of which are
+ok:
+
+1. A table entry was written after `tWrite` but before `tScan`, leaving behind
+   a corresponding index entry. This case is safe since we're just repeating
+   previous work.
+2. A backfiller dies before it checkpoints, leaving behind an index entry. When
+   the backfiller restarts, it encounters its previously written entry. This
+   case is safe since we're again just repeating previous work.
+
+There's never a case where this would accidentally cause a uniqueness
+constraint violation, since the index entry contains the (guaranteed unique)
+primary key of the table in the value. A uniqueness constraint violation always
+manifests as an unexpected value for a key.
+
+For example, when the backfiller encounters the pair `2` `b`, the corresponding
+index entry is `b` `2`, so the backfiller does nothing and moves on.
+
+## The `InitPut` fails
+
+In all other cases, the `InitPut` will fail. In these cases, we will retry the
+operation transactionally with a fresh read to determine whether to return a
+uniqueness constraint violation or to write a more up to date value.
```

Comment:
We do not need to write a more up to date value since the entire cluster has already converged to the WRITE_AND_DELETE_ONLY state. This is done only to return a uniqueness constraint violation and to update keys within the batch that were never written. If we are updating a 100 keys, if one key is problematic we would hit an error for the entire batch. We could try another approach:

1. Not return an error from the InitPut so that unchanged keys get outputted, while others get ignored.
2. Run select count(*) on the table and on the index at an AS OF SYSTEM TIMESTAMP to see if they match up at the end of the backfill.

With the above approach we would never need to run the second transaction, with the tradeoff that we insert the verification step at the end, but one can argue we in any case should verify (will be adding this later, so why not now)